polar-core/src/filter.rs

```diff
@@ -0,0 +1,540 @@
+            }
+        });
+
+        // add every variable whose type we know to the entities map.
+        //
+        // partition variables into equivalence classes
+        crate::data_filtering::partition_equivs(equivs)
+            // map each variable to its own equivalence class
+            .into_iter()
+            .map(std::rc::Rc::new)
+            .flat_map(|cls| {
+                cls.iter()
+                    .cloned()
+                    .map(|pv| (pv, cls.clone()))
+                    .collect::<Vec<_>>()
+            })
+            // for each variable k, if a variable in k's
+            // eq class has a known type, then assign that
+            // type to k.
+            .filter_map(|(k, v)| {
+                v.iter()
+                    .find_map(|eq| self.get_type(eq.clone()))
+                    .map(|t| (k, t))
+            })
+            .collect::<Vec<_>>() // so the closure ^^^ lets go of &mut self
+            .into_iter()
+            .for_each(|(k, t)| {
+                // add them to the entities map
+                self.entities.insert(k, t);
+            });
+
+        // every variable that needs a type
+        // should now hopefully have a type.
+        // now add a condition for each partial.
+        // this also populates the relations.
+        for op in ops {
+            self.add_constraint(op)?;
+        }
+
+        Ok(self)
+    }
+
+    fn build_filter(
+        type_info: TypeInfo,
+        parts: Vec<Operation>,
+        class: &str,
+    ) -> FilterResult<Filter> {
+        // turn an isa from the partial results into a pathvar -> type pair
+        fn isa2entity(op: Operation) -> Option<FilterResult<(PathVar, TypeName)>> {
+            match op.args[1].value() {
+                Value::Pattern(Pattern::Instance(InstanceLiteral { tag, fields }))
+                    if fields.is_empty() =>
+                {
+                    // FIXME(gw) this is to work around a complicated simplifier
+                    // bug that causes external instances to sometimes be embedded
+                    // in partials. term2pathvar will fail if the base of the path
+                    // is an external instance and not a var. we can't check if the
+                    // isa is true from in here, so just drop it for now. the host
+                    // should check for these before building the filter.
+                    match PathVar::from_term(&op.args[0]) {
+                        Err(_) => None,
+                        Ok(p) => Some(Ok((p, tag.0.clone()))),
+                    }
+                }
+                _ => Some(unsupported_op_error(op)),
+            }
+        }
+
+        // we use isa constraints to initialize the entities map
+        let (isas, othas): (Set<_>, Set<_>) = parts
+            .into_iter()
+            .partition(|op| op.operator == Operator::Isa);
+
+        // entities maps variable paths to types
+        let entities = isas
```

Comment:
This assumes that all the `isas` are satisfied without checking. i.e. that `x matches Foo` is true, or that `x.bar matches Bar` is true.

Is this assumption always correct? (I think so because we check with `IsaWithPath`?)

Reply:
These are partials from the simplifier, so Polar thinks they're true. If they aren't then we have bigger problems 🤷‍♀️

Reply:
That's not really true though, they need to be checked still.

When the simplifier outputs `x.y = 1`, it doesn't think it's true, it's just cannot prove that it's false. Same for `x matches Y`. We check `x.y = 1` in the host by translating to `FROM x WHERE y = 1`. In previous adapters we've done the same for `matches` constraints by checking them in the host library: https://github.com/osohq/oso/blob/cb8e7dc40daf2b47a0fcce59c3026cc072c2989b/languages/python/sqlalchemy-oso/sqlalchemy_oso/partial.py#L193

I'm not sure if that's needed here.

Reply:
I would need to think about this more to come up with a better example of how this could cause an issue.

Reply:
> When the simplifier outputs x.y = 1, it doesn't think it's true, it's just cannot prove that it's false.

I don't think that's accurate. It's something that must be true for the partial to hold for a given object.

Reply:
I'm curious what it could be